primary/src/block_synchronizer/handler.rs

```diff
@@ -0,0 +1,249 @@
+/// functionality in a synchronous way without having to deal with message
+/// emission. The BlockSynchronizer on its own for the certificates that
+/// fetches on the fly from peers doesn't care/deal about any other validation
+/// checks than the basic verification offered via the certificate entity
+/// it self. For that reason the Handler offers methods to submit the fetched
+/// from peers certificates for further validation & processing (e.x ensure
+/// parents history is causally complete) to the core and wait until the
+/// certificate has been processed, before it returns it back as result.
+#[automock]
+#[async_trait]
+pub trait Handler<PublicKey: VerifyingKey> {
+    /// It retrieves the requested blocks via the block_synchronizer making
+    /// sure though that they are fully validated. The certificates will only
+    /// be returned when they have properly processed via the core module
+    /// and made sure all the requirements have been fulfilled.
+    async fn get_and_synchronize_block_headers(
+        &self,
+        block_ids: Vec<CertificateDigest>,
+    ) -> Vec<Result<Certificate<PublicKey>, Error>>;
+
+    /// It retrieves the requested blocks via the block_synchronizer, but it
+    /// doesn't synchronize the fetched headers, meaning that no processing
+    /// will take place (causal completion etc).
+    async fn get_block_headers(
+        &self,
+        block_ids: Vec<CertificateDigest>,
+    ) -> Vec<BlockSynchronizeResult<BlockHeader<PublicKey>>>;
+}
+
+/// A helper struct to allow us access the block_synchronizer in a synchronous
+/// way. It also offers methods to both fetch the certificates and way to
+/// process them and causally complete their history.
+pub struct BlockSynchronizerHandler<PublicKey: VerifyingKey> {
+    /// Channel to send commands to the block_synchronizer.
+    tx_block_synchronizer: Sender<Command<PublicKey>>,
+
+    /// Channel to send the fetched certificates to Core for
+    /// further processing, validation and possibly causal
+    /// completion.
+    tx_core: Sender<PrimaryMessage<PublicKey>>,
+
+    /// The store that holds the certificates.
+    certificate_store: Store<CertificateDigest, Certificate<PublicKey>>,
+
+    /// The timeout while waiting for a certificate to become available
+    /// after submitting for processing to core.
+    certificate_deliver_timeout: Duration,
+}
+
+impl<PublicKey: VerifyingKey> BlockSynchronizerHandler<PublicKey> {
+    pub fn new(
+        tx_block_synchronizer: Sender<Command<PublicKey>>,
+        tx_core: Sender<PrimaryMessage<PublicKey>>,
+        certificate_store: Store<CertificateDigest, Certificate<PublicKey>>,
+        certificate_deliver_timeout: Duration,
+    ) -> Self {
+        Self {
+            tx_block_synchronizer,
+            tx_core,
+            certificate_store,
+            certificate_deliver_timeout,
+        }
+    }
+
+    #[instrument(level = "debug", skip_all)]
```

Comment:
Ah, in order for this to work the method needs to return a `Result<T, E>` , otherwise the macro won't be able to capture the errors - the method returns a `Vec` of results. I'll log the errors on higher level